### PR TITLE
Wrap producer in metrics

### PR DESCRIPTION
**Pull request by KensoDev**
[https://globality.atlassian.net/browse/DEVOPS-636](https://globality.atlassian.net/browse/DEVOPS-636)

### CODE REVIEW

File: microcosm_pubsub/producer.py [hunk 1]
```diff
@@ -100,16 +110,32 @@ def publish_message(self, pubsub_message: PubsubMessage):
...
-                Message=pubsub_message.message,
-                MessageAttributes=pubsub_message.message_attributes,
-            )
+            try:
+                result = self.sns_client.publish(
+                    TopicArn=pubsub_message.topic_arn,
+                    Message=pubsub_message.message,
+                    MessageAttributes=pubsub_message.message_attributes,
+                )
+            except Exception as e:
```
Comment (KensoDev) [inline microcosm_pubsub/producer.py]: Tracking the botocore exception here (after the timeout is done).
We catch it here but throw it later after we send the stats

File: microcosm_pubsub/producer.py [hunk 2]
```diff
@@ -100,16 +110,32 @@ def publish_message(self, pubsub_message: PubsubMessage):
             **pubsub_message.opaque_data
         )
         self.logger.debug("Publishing message with media type {media_type}", extra=extra)
+        publish_result = "SUCCESS"
```
Comment (KensoDev) [inline microcosm_pubsub/producer.py]: This is sort of iffy, I know.

File: microcosm_pubsub/producer.py [hunk 3]
```diff
@@ -100,16 +110,32 @@ def publish_message(self, pubsub_message: PubsubMessage):
...
+                publish_exception = e
+                publish_result = "FAILURE"
+
+        self.pubsub_producer_metrics(
+            publish_result=publish_result,
+            **extra
+        )
+
+        if publish_exception:
+            raise Exception(f"Could not publish message, SNS producer error")
```
Comment (GabeLambda) [inline microcosm_pubsub/producer.py]: Why not just re-raise the error?

File: microcosm_pubsub/metrics.py [hunk 4]
```diff
@@ -120,3 +120,52 @@ def __call__(self, elapsed_time: float, message_batch_size: int):
...
+    Send metrics regarding the producer
+
+    """
+
+    def __init__(self, graph):
+        self.metrics = self.get_metrics(graph)
+        self.enabled = bool(
+            self.metrics
+            and self.metrics.host != "localhost"
+            and graph.config.pubsub_send_metrics.enabled
```
Comment (pankajghosh) [inline microcosm_pubsub/metrics.py]: Should this be pubsub_producer_metrics.enabled?

Comment (KensoDev) [reply microcosm_pubsub/metrics.py]: No. This is the generic one that we send in ansible.
I don't want to have a fine grained one. If we track pubsub metrics, it should track dispatcher and producer.

File: microcosm_pubsub/producer.py (hunk 3 above)
Comment (KensoDev) [reply microcosm_pubsub/producer.py]: No reason outside of logging search

File: microcosm_pubsub/metrics.py (hunk 4 above)
Comment (pankajghosh) [reply microcosm_pubsub/metrics.py]: Sounds good.

File: microcosm_pubsub/producer.py (hunk 3 above)
Comment (GabeLambda) [reply microcosm_pubsub/producer.py]: Then why not:
```
             raise Exception(f"Could not publish message, SNS producer error: {publish_exception}")
```
I for one would like to know what the actual problem was.

File: microcosm_pubsub/metrics.py [hunk 5]
```diff
@@ -120,3 +120,52 @@ def __call__(self, elapsed_time: float, message_batch_size: int):
...
+            return
+
+        tags = [
+            "source:microcosm-pubsub",
+            f"publish_result:{publish_result}",
+        ]
+
+        self.metrics.histogram(
+            "sns_producer",
+            elapsed_time,
```
Comment (GabeLambda) [inline microcosm_pubsub/metrics.py]: Nit: convert to ms?

File: microcosm_pubsub/producer.py (hunk 3 above)
Comment (KensoDev) [reply microcosm_pubsub/producer.py]: Done!

**Comment by GabeLambda** [comment]
Is there a reason this couldn't be rolled into the original publish metrics?

I'm wondering if we couldn't record the metrics here:
https://github.com/globality-corp/microcosm-pubsub/blob/89b118590d052433a2f3175a64e03c47e26121b2/microcosm_pubsub/result.py#L71

https://github.com/globality-corp/microcosm-pubsub/blob/develop/microcosm_pubsub/metrics.py#L37

That probably does end up adding a bit more work, at the benefit of having less exception handling.

**Pull request merged**